### PR TITLE
Load libssh2 at runtime

### PR DESCRIPTION
**Pull request by carlosmn**
This is a continuation of #3785 which takes it further by removing the attempt to detect and link against libssh2 at build time. Once we go run-time we should own that.

There's an aspect which is not yet quite complete, namely which SOVERSIONs we should look up. The Linux version should probably go like the Darwin one and use simply "1" but we'll see what CI says.

I've tested that loading the library works when placed alongside the clar binary, but we still need to figure out what to do for testing on Windows.

### CODE REVIEW

File: src/transports/ssh.c [hunk 1]
```diff
@@ -893,17 +978,101 @@ int git_transport_ssh_with_paths(git_transport **out, git_remote *owner, void *p
...
+};
+
+/**
+ * Try to load a supported version of libssh2
+ */
+static void load_libssh2(void)
+{
+#ifdef GIT_WIN32
+	/* Windows doesn't do SOVERSIONs but you ship your own library so we're probably OK */
+	git_libssh2_handle = LoadLibrary("libssh2.dll");
```
Comment (tkelman) [inline src/transports/ssh.c]: some environments (msys2 probably, will have to check) approximate sonames via libssh2-#.dll

**Comment by carlosmn** [comment]
This opens the question of what the SSH capability means. Does it mean we'd support ssh if we could find libssh2 or does it mean we've found it and can do ssh?

**Comment by tiennou** [comment]
From the top of my head, I'd prefer that it meant we've found it and can do ssh. IMHO the other case is meaningful only when we allow the application to provide its own libssh2.

Alternatively, split the `GIT_FEATURE_SSH` into `GIT_FEATURE_SSH_ENABLED`/`GIT_FEATURE_SSH_AVAILABLE` ?

**Comment by carlosmn** [comment]
>From the top of my head, I'd prefer that it meant we've found it and can do ssh

Yes, that make sense, but we've been using this as a way to see the compile-time options so it'd be a change in function. It might end up being the better solution, though.

I've tried to get this working on Appveyor but can't reproduce the success from my VM. I've uploaded the same dll I'm using, but neither the 32-bit or 64-bit versions seem to want to load them.

**Comment by ethomson** [comment]
> I've tried to get this working on Appveyor but can't reproduce the success from my VM. I've uploaded the same dll I'm using, but neither the 32-bit or 64-bit versions seem to want to load them.

Naive guess:  is libssh2 built dynamically with a different version of the C runtime?  Can you load it (on your machine) with dependency walker?

Are you building libssh2 with `/MD`?  Can you instead build with `/MT`?

**Comment by carlosmn** [comment]
💃  once I built them statically (so they don't depend at runtime on a newer VS), we're able to find `libssh2.dll`.

The mingw build is still failing because apparently they don't have the unix library loading functionality headers; hopefully we can just use the Win32 ones.

**Comment by carlosmn** [comment]
This is now passing al-round, so "all" that's left is to make a comprehensible commit history and adjust the documentation to mention the DLL loading.

**Comment by jacquesg** [comment]
Does this mean that we can longer link libssh2 statically?

**Comment by carlosmn** [comment]
With the code as-is, it wouldn't support a statically-linked libssh2 but you can load symbols in the same object so it could with small changes.

**Comment by jacquesg** [comment]
That would be great as it is often more convenient to statically link dependencies if their licenses allow for it.

File: src/transports/ssh.c [hunk 2]
```diff
@@ -15,10 +19,75 @@
...
+
+#define LOOKUP_LIBSSH2_SYMBOL(symbol) \
+	git_ ## symbol = (void *) GetProcAddress(git_libssh2_handle, #symbol)
+
+#else
+
+static void *git_libssh2_handle;
+
+#define LOOKUP_LIBSSH2_SYMBOL(symbol) \
+	git_ ## symbol = dlsym(git_libssh2_handle, #symbol)
```
Comment (ethomson) [inline src/transports/ssh.c]: Need error checking here.

File: src/transports/ssh.c [hunk 3]
```diff
@@ -15,10 +19,75 @@
...
+
+/* These macros allow us to refer to libssh2 symbols while
+ * supporting dynamic linking.
+ */
+#ifdef GIT_SSH_LOADLIBRARY
+
+static HMODULE git_libssh2_handle;
+
+#define LOOKUP_LIBSSH2_SYMBOL(symbol) \
+	git_ ## symbol = (void *) GetProcAddress(git_libssh2_handle, #symbol)
```
Comment (ethomson) [inline src/transports/ssh.c]: Need error checking here.

**Comment by ethomson** [comment]
Yeah, I agree with @jacquesg here - it would be nice if we could allow for static linking and then just `#define git_libssh2_whatever libssh2_whatever`.  I actually think that static linking should be the general case for ssh support and this dynamic loading should be explicitly opt-in.

**Comment by carlosmn** [comment]
I've added the error handling for not finding the required functions. I've also re-enabled finding the library at runtime, though I'm not sure that adds a lot considering `-lcurl` nowadays implies `-lssh2` as well (that was fun to try to piece together).

So we now have `-DUSE_SSH=ON` and `-DUSE_SSH_RUNTIME=ON`, the latter of which will enable runtime detection. In order to keep code similar, we still load them with `dlsym` at runtime, but look at the current image instead of trying to load a library. I haven't tested that one out too much yet, but the manpage leads me to believe that it should work in that situation.

**Comment by carlosmn** [comment]
And of course `RTLD_DEFAULT` is actually `NULL` in many platforms, so we can't distinguish those values.

**Comment by carlosmn** [comment]
I'm hitting a wall trying to get the same codebase to target all three forms of linking. The code here seems to be unable to properly load symbols which are already available and is making unrelated code crash.

I've spent all day today trying to approach this via weak linking, which was my hope for a clean implementation with overridable functions, but glibc's loader is trying to be efficient and won't let libssh2 override the weak symbols in `libgit2.so` or `libgit2_clar`, so that's not going to go anywhere.

**Comment by seltzy** [comment]
Why does this remove the `EMBED_SSH_PATH` option? It seemed like a more convenient option, especially on toolchains/systems that don't include `pkg-config`. In particular, this seems like it would make cross-compilation more difficult. I already have enough trouble using MinGW on Linux to target Windows as also noted in #3926 - in my case it ends up using the `PC` definitions to get `gcrypt` and `libssh2` even with `EMBED_SSH_PATH` which subsequently causes builds to fail at link-time since there's no shared libssh2 built with the same toolchain. Will runtime loading mitigate this, or make things more confusing?

In my case, I have a project where I mostly build my dependencies from source as this yields more consistent building behaviour across target platforms and has helped reduced maintenance for me. Embedding libssh2 is quite nice for this, but perhaps it's not The Right Way™. This concern could be fueled by inexperience or lack of understanding, but I would like to better understand the implications of this for Windows builds and also for cross-compiling.

**Comment by carlosmn** [comment]
After some further discussion elsewhere, I've come to the conclusion that this isn't a worthwhile time investment.

The primary driver for this was to allow libssh2 to be picked up dynamically on Windows while keeping the "normal" dynamic linking in the other platforms, where the normal course of business lets us detect and support libssh2.

But in order for this to work properly, we'd still have to provide scripts or instructions for people who are not used to the C world on how to build it. At that point, it'd be about as much effort on that side, and significantly less on this side, to allow them to build libgit2 with a built-in libssh2 (though redistribution would still be an unsolved question).

**Pull request closed without merging**